apis/oam/v1alpha2/core_workload_types.go

```diff
@@ -0,0 +1,398 @@
+	Port int32 `json:"containerPort"`
+
+	// TODO(negz): Use +kubebuilder:default marker to default Protocol to TCP
+	// once we're generating v1 CRDs.
+
+	// Protocol used by the server listening on this port.
+	// +kubebuilder:validation:Enum=TCP;UDP
+	// +optional
+	Protocol *TransportProtocol `json:"protocol,omitempty"`
+}
+
+// An ExecProbe probes a container's health by executing a command.
+type ExecProbe struct {
+	// Command to be run by this probe.
+	Command []string `json:"command"`
+}
+
+// A HTTPHeader to be passed when probing a container.
+type HTTPHeader struct {
+	// Name of this HTTP header. Must be unique per probe.
+	Name string `json:"name"`
+
+	// Value of this HTTP header.
+	Value string `json:"value"`
+}
+
+// A HTTPGetProbe probes a container's health by sending an HTTP GET request.
+type HTTPGetProbe struct {
+	// Path to probe, e.g. '/healthz'.
+	Path string `json:"path"`
+
+	// Port to probe.
+	Port int32 `json:"port"`
+
+	// HTTPHeaders to send with the GET request.
+	// +optional
+	HTTPHeaders []HTTPHeader `json:"httpHeaders,omitempty"`
+}
+
+// A TCPSocketProbe probes a container's health by connecting to a TCP socket.
+type TCPSocketProbe struct {
+	// Port this probe should connect to.
+	Port int32 `json:"port"`
+}
+
+// A ContainerHealthProbe specifies how to probe the health of a container.
+// Exactly one of Exec, HTTPGet, or TCPSocket must be specified.
+type ContainerHealthProbe struct {
+	// Exec probes a container's health by executing a command.
+	// +optional
+	Exec *ExecProbe `json:"exec,omitempty"`
+
+	// HTTPGet probes a container's health by sending an HTTP GET request.
+	// +optional
+	HTTPGet *HTTPGetProbe `json:"httpGet,omitempty"`
+
+	// TCPSocketProbe probes a container's health by connecting to a TCP socket.
+	// +optional
+	TCPSocket *TCPSocketProbe `json:"tcpSocket,omitempty"`
+
+	// InitialDelaySeconds after a container starts before the first probe.
+	// +optional
+	InitialDelaySeconds *int32 `json:"initialDelaySeconds,omitempty"`
+
+	// TODO(negz): Use +kubebuilder:default marker to default PeriodSeconds,
+	// TimeoutSeconds, SuccessThreshold, and FailureThreshold to 10, 1, 1, and 3
+	// respectively once we're generating v1 CRDs.
+
+	// PeriodSeconds between probes.
+	// +optional
+	PeriodSeconds *int32 `json:"periodSeconds,omitempty"`
+
+	// TimeoutSeconds after which the probe times out.
+	// +optional
+	TimeoutSeconds *int32 `json:"timeoutSeconds,omitempty"`
+
+	// SuccessThreshold specifies how many consecutive probes must success in
+	// order for the container to be considered healthy.
+	// +optional
+	SuccessThreshold *int32 `json:"successThreshold,omitempty"`
+
+	// FailureThreshold specifies how many consecutive probes must fail in order
+	// for the container to be considered healthy.
+	// +optional
+	FailureThreshold *int32 `json:"failureThreshold,omitempty"`
+}
+
+// A Container represents an Open Containers Initiative (OCI) container.
+type Container struct {
```

Comment:
Where did we land on using Kubernetes `Container` type directly? cc @ryanzhang-oss

Reply:
Unclear at this stage - I've raised https://github.com/oam-dev/spec/issues/325 to track and discuss this.